README.md
opentibia-xml-data-validator
============================

xml data files validator for opentibia servers based on xml files


DOC/changelog
XML Validator Changelog.
Still in beta phase, it works ok, but several features are not  implemented yet.

0.0.5a (beta)

-fixed chmod function
-fixed some issues with UNIX systems
-added automatic change possibility for premDays
-automatic change possibility to account type

0.0.5 (beta)

-fixed parseFile (third argument depracated)
-exceptions added (files you dont want to check)
-saving raport to HTML (configurated)
-polish language file added
-experience validation
-lookdir validation

0.0.4 (beta)

-fixed parseFile function 
-small optimalization
-checking php version if capable
-script changes max_execution_time if safe mode is disabled
-sex validation
-enhanced premDays check
-parseFile function rebuilded
-possibility to delete files with missing account, or accounts with missing characters ($unlink)
-rebuilding error_color function
-fixed DOMDocument error
-user error filter added
-support for php4 has been disabled
-fixed bug with counting examined files
-multilanguage support
-script structure changed
-some small errors fixed

0.0.3 (not stable)

-ajax support enabled
-file couner function
-script output fixed
-script optymalization up to 30%

0.0.2 (not stable)

-fixed bug with is_numeric
-new xml features examined
-first functions used

0.0.1 (not stable)

-xml sex axamination
-xml voc examination
-xml maglev examination
-first release




index.php
<html>  
<head> 
<!-- Oink! oink! -->
<!-- XML Server Validator by Pavlus -->
<script src="js/jquery.js"></script>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>
<title>Validator XML</title> 


</head> 
<?php

require_once ('config.php');

//preloading language file


if(file_exists('lang/'.$lang.'.php'))
	{
		require_once('lang/'.$lang.'.php');
		}
			else
			{
				echo 'Wrong language type set in config.php, $lang variable!';
				exit();
				}
				
	//checking core status
if(filesize('validation.php') != CORE_SIZE)
	{
		echo '<font color="red"><b>'.$misc['core_changed'].'</font></b><br>';
		}
				?>
<body>  

<script language="javascript">  
 
jQuery(function($) {  
$("#contentArea").load("validation.php");  
});  
  
$().ajaxSend(function(r,s){  
$("#contentLoading").show();  
});  
  
$().ajaxStop(function(r,s){  
$("#contentLoading").fadeOut("slow");  
});  
   
</script> 

<center>

<div id="contentLoading" class="contentLoading"> 
<?php echo $gener; ?><br>
<img src="img/ajax-loader.gif" alt="Loading report, please wait..."><br>
<small><?php echo $eta; ?></small>
</div>  
</center>
<div id="contentArea">  
</div>  
</body>  
</html>  

config.php
<?php
/*
====+++========
Open Tibia XML Data Validator
Under GPL v3 License
Author: Pavlus
Configuration file
====+++========
*/

#################### STANDARD SETTINGS #######################
$lang = 'en'; // choose your language 'pl' for polish, and 'en' - english
$path['data'] = 'G:/OTS/data/'; // path to your opentibia server data directory
$saveLog = 1; //0 - no raport / 1 - default html export / 2 - export to file
$account['digits'] = 6;	//number of digits for account
$premDay['limit'] = 300; // if player have more than $premDay will show error (yurots new year bug)

#################### FILE WRITE/REMOVE SETTINGS #######################
$change = 0; //set this to "1" if u want to allow modifying files by this script
$premDay['rigid'] = 1; //if modifying is allowed, script will set 1 premday when player exceed the limit of premdays 
$unlink = 0; // set this to 0 if you dont want validator delete empty accounts or missing player account
//remember to make a BACKUP before running script!

######################## EXCEPTIONS #########################
$except['acc'] = array('0.xml', '1.xml'); //if you doesn`t want to examine some account numbers, divide by coma.
$except['pla']= array('Marco Polo.xml', 'GM.xml'); //the same exception for players

#END OF CONFIGURATION




/*
DO NOT EDIT CODE BELOW! UNDER YOU KNOW WHAT ARE YOU DOING
##########################################################
*/

//constants

define ("VERSION", "0.0.5a");
define("STATUS", "Beta");
define("CORE_SIZE", "11891");

//OS check

if (strtoupper(substr(PHP_OS, 0, 3)) === 'WIN') {
    $UNIX = 0;
} else {
    $UNIX = 1;
}

//generate players, accounts, vip folder

$path['acc'] = $path['data'].'/accounts/'; // accounts folder path f.e /home/pavlus/ots/accounts/
$path['pla'] = $path['data'].'/players/'; // players path
$path['vip'] = $path['data'].'/vip/'; //vip path

//functions


function parseFile($path, $echo)
	{
		
		if($echo == 1)
		
		{
		
		$n =  end(explode('/', $path));
			echo '<b>'.$n.'</b>';
			}
				
		
			else if($echo == 0)
				{
						$n = end(explode ('/', $path));
								return $n;
						}
		}
		

		
function error_color($string, $color)
	{
		echo '<b><font color='.$color.'>'.$string.'</font></b>';
		
		}
		
		
function count_files($mode)
{
	global $path;
		global $pla_num;
			global $acc_num;
	
	if($mode == 1)
		{
				$od = openDir($path['pla']);

					$pla_num = -2;
						while(readDir($od) !== false){
							$pla_num++;
		}
					closeDir($od);

			return $pla_num;
}

if($mode == 0)
		{
				$od = openDir($path['acc']);

						$acc_num = -2;
							while(readDir($od) !== false){
								$acc_num++;
		}
					closeDir($od);
		return $acc_num; 	
	
	
	}
}

function gentime() {
    static $a;
    if($a == 0) $a = microtime(true);
    else return (string)(microtime(true)-$a);
} 

	
function listErrors()
{
	global $err;
	
	if($err > 1)
		{
			echo error_color($err, red);

			}
			
			else if ($err == 1)
				{
					echo error_color('1', red);
					}
				else if ($err == 0)
				{
					echo '0';
					}
	}
	
	function deletedFiles()
	{
		global $unlinked;
		global $unlink;
			
			if($unlink == 1)
			{
				if($unlinked > 1)
				{
				echo error_color($unlinked, green);
				}
					else if($unlinked == 0)
					{
						echo error_color('0', green);
						}
				}
	}
	
	function changedFiles()
	{
		global $changed;
		global $change;
			
			if($change == 1)
			{
				if($changed > 1)
				{
				echo error_color($changed, orange);
				}
					else if($changed == 0)
					{
						echo error_color('0', orange);
						}
				}
	}
			

?>


lang/en.php
<?php
/*
====+++========
Open Tibia XML Data Validator
Under GPL v3 License
Author: Pavlus
Language file
====+++========
*/

require_once('./config.php'); //this line must be included!

//error messages

$error['php_version'] = 'Your PHP version is out of date! At least 5.0.0 version is required. Script is terminating';
$error['directory'] = 'You have entered bad path to your account, players or vip catalogue. Change this in config.php file. Script is now terminating.';
$error['simplexml'] = 'Your PHP doesn`t support simplexml_file_load function. Script is terminating.';
$error['unlink'] = 'Your PHP doesn`t support unlink function. Script is terminating.';
$error['xml_syntax'] = 'Fatal error: Syntax broken in XML file!';
$error['wrong_digits'] = 'Account number not equals '.$account['digits'].'';
$error['empty_pass'] = 'Empty password';
$error['wrong_type'] = 'Wrong account type';
$error['prem_without_value'] = 'remDays attribute is empty';
$error['prem_not_int'] = 'premDays is not numeric';
$error['prem_less_zero'] = 'premDays is less than zero!';
$error['prem_max'] = 'premDays is bigger than '.$premDay['limit'].'';
$error['empty_account'] = 'Account file without character in it';
$error['account_max'] = 'Maximum characters for account exceeded ('.$maxchars.')';
$error['account_miss'] = 'Account file missing for player';
$error['sum'] = 'Number of errors:';
$error['deleting'] = 'Error occured during deletion your folder with files must have write-acces rights!<br>';
$error['unlinked'] = 'Number of deleted files:';
$error['changed'] = 'Number of changed files:';
$error['sex_numeric'] = 'Sex attriubute is not numerical value';
$error['sex_value'] = 'Sex attribute have not permitted value';
$error['lookdir_value'] = 'Lookdir attribute have not permitted value';
$error['exp_value'] = 'Exp attribute have not permitted value';
$error['vip_exists'] = 'Missing account file for this VIP';
$error['health'] = 'Health now and Health maximum have undesired value';
$error['mana'] = 'Mana now and Mana maximum have undesired value';  
$error['manamax_l'] = 'Mana max not corresponds to level'; 
$error['healthmax_l'] = 'Health max not corresponds to level'; 
$error['cap_l'] = 'Capacity not corresponds to level'; 


//header msg

$head['title'] = 'XML VALIDATOR RAPORT';
$head['version'] = 'VERSION';
$head['date'] = 'DONE';
$head['file_count'] = 'EXAMINED FILES';


//raport output titles

$title['vip'] = 'Checking vip files: OK';
$title['exp'] = 'Checking experience: OK';
$title['lookdir'] = 'Checking lookdir: OK';
$title['sex_dig'] = 'Checking sex: OK'; 
$title['digits'] = 'Checking number of account digits: OK';
$title['pass'] = 'Checking passwords: OK';
$title['types'] = 'Checking types: OK';
$title['prem'] = 'Checking premDays: OK';
$title['char'] = 'Checking characters: OK';
$title['acc_num'] = 'Checking account number: OK';
$title['health'] = 'Checking players health: OK';
$title['mana'] = 'Checking players mana: OK';   
$title['var_l'] = 'Level versus other player data: OK';

//head messages, what kind of files we examined

$mode['accounts'] = 'Results of checking accounts folder';
$mode['players'] = 'Results of checking players folder';

//misc

$misc['core_changed'] = 'Core file of Validator has been changed! You run now at your own risk';
$misc['raport_chmod'] = 'LOG folder should have read & write acces! (chmod 777)';
$sec = 'sec';
$gener = 'Raport is generating now, please be patient';
$eta = 'It may take few seconds or few minutes, it depends of number of files to be examined';



?>


lang/pl.php
<?php
/*
====+++========
Open Tibia XML Data Validator
Under GPL v3 License
Author: Pavlus
Language file
====+++========
*/

require_once('./config.php'); //this line must be included!

//error messages

$error['php_version'] = 'Twoja wersja PHP jest przestarzała, wymagane jest co najmniej PHP w wersji > 5.0.0';
$error['directory'] = 'Ścieżka do folderów account, players lub vip jest niepoprawna, musisz zmienić to w config.php.';
$error['simplexml'] = 'Funkcja simplexml_load_file niedostępna, sprawdź php.ini';
$error['unlink'] = 'Twoje PHP nie supportuje funkcji unlink, zmień ustawienia w php.ini.';
$error['xml_syntax'] = 'Błąd krytyczny, składnia XML niepoprawna!';
$error['wrong_digits'] = 'Liczba cyfr w numerze konta inna niż: '.$account['digits'].'';
$error['empty_pass'] = 'Puste hasło';
$error['wrong_type'] = 'Zły account type';
$error['prem_without_value'] = 'Atrybut premDays nie zawiera żadnej wartości';
$error['prem_not_int'] = 'Atrybut premDays nie jest wartością numeryczną';
$error['prem_less_zero'] = 'Atrybut premDays jest mniejszy od zera!';
$error['prem_max'] = 'Liczba dni premium jest większa niż: '.$premDay['limit'].'';
$error['empty_account'] = 'Plik konta nie zawiera postaci';
$error['account_max'] = 'Maksymalna liczba postaci na koncie: ('.$maxchars.') przekroczona';
$error['account_miss'] = 'Plik account nie istnieje dla tego gracza';
$error['sum'] = 'Liczba błędów:';
$error['deleting'] = 'Błąd podczas usuwania, musisz mieć prawa odczytu i zapisu!<br>';
$error['unlinked'] = 'Liczba usuniętych plików:';
$error['changed'] ='Liczba zmienionych plików:';
$error['sex_numeric'] = 'Atrybut sex nie jest wartością numeryczną';
$error['sex_value'] = 'Atrybut sex ma nieprawidłową wartość';
$error['lookdir_value'] = 'Atrybut lookdir ma nieprawidłową wartość';
$error['exp_value'] = 'Atrybut exp ma nieprawidłową wartość';
$error['vip_exists'] = 'Plik viplisty dla tego konta, nie powinien istnieć';

//header msg

$head['title'] = 'XML VALIDATOR RAPORT';
$head['version'] = 'WERSJA';
$head['date'] = 'WYKONANO';
$head['file_count'] = 'PRZESKANOWANYCH PLIKÓW';


//raport output titles

$title['vip'] = 'Sprawdzanie viplisty: OK';
$title['exp'] = 'Sprawdzanie experience: OK';
$title['lookdir'] = 'Sprawdzanie lookdir: OK';
$title['sex_dig'] = 'Sprawdzanie sex: OK'; 
$title['digits'] = 'Sprawdzanie liczb numeru konta: OK';
$title['pass'] = 'Sprawdzanie haseł: OK';
$title['types'] = 'Sprawdzanie typów: OK';
$title['prem'] = 'Sprawdzanie premDays: OK';
$title['char'] = 'Sprawdzanie postaci: OK';
$title['acc_num'] = 'Sprawdzanie numerów kont: OK';

//head messages, what kind of files we examined

$mode['accounts'] = 'Wyniki sprawdzania folderu accounts';
$mode['players'] = 'Wyniki sprawdzania folderu players';

//misc

$misc['core_changed'] = 'Główny plik Validatora został zmieniony! Uruchamiasz go teraz na własną odpowiedzialność';
$misc['raport_chmod'] = 'Katalog LOG powinien miec możliwość odczytu i zapisu przez właściciela (chmod 777)';
$sec = 'sek';
$gener = 'Trwa generowanie raportu, proszę czekać';
$eta = 'Test może potrwać od kilku sekund do kilku minut, w zależności od liczby plików do przeskanowania';



?>


validation.php
<?php

/*
====+++========
Open Tibia XML Data Validator
Under GPL v3 License
Author: Pavlus
Validator main file
====+++========
*/


require_once ('config.php');

//loading language file
if(file_exists('lang/'.$lang.'.php'))
	{
		require_once('lang/'.$lang.'.php');
		}
			else
			{
				echo 'Wrong language type. Change in config.php: $lang.';
				exit();
				}

//php version check
if(version_compare(PHP_VERSION, '5.0.0', '<'))
	{
		echo $error['php_version'].' </br>';
		exit();
		}
		
//checking safe mode
if(!ini_get('safe_mode'))
	{
//checking timeout ini file
$timeout = ini_get('max_execution_time'); //ini file check
if($timeout < 540)
{
$set_timeout = @ini_set('max_execution_time',0);

//	if(!$set_timeout)
	//	echo 'FAILED TO CHANGE PHP.INI';
}

	}
			
//user error finder
//will not work if this function is disabled
//what i wanted to say by writing this? dunno....

if(!function_exists(simplexml_load_file))
	{
		echo $error['simplexml'];
		exit();
		}	
if(!function_exists(unlink))
	{
		echo $error['unlink'];
		exit();
		}
//path is incorrect?
if(!is_dir($path['acc']) || !is_dir($path['pla']) || !is_dir($path['vip']))
	{
		echo $error['directory'];
		exit();
		}
		
/*
Main code
*/
ob_start();

if($UNIX == 1)
	{
		
		$old_chmod = substr(sprintf('%o', fileperms($path['pla'])), -4); //retriving chmod			
				$old_chmod_changed = octdec($old_chmod); //we have to do it :/		
		@chmod($path['pla'], 0777);
			//we shall change chmods of our working directories before unlinking some files
			@chmod($path['acc'], 0777);
				@chmod($path['vip'], 0777);
			}

		count_files(0);
		count_files(1); // examined file counter
	
$date = date("F j, Y, g:i a");
$validated = $pla_num + $acc_num;
$excluded = $files.'0.xml';

//raport header



echo '==============================<br>
= '.$head['title'].'<br>
= '.$head['version'].': '.VERSION.' ('.STATUS.')<br>
= '.$head['date'].': '.$date.'<br>
= '.$head['file_count'].': '.$validated.'<br>
          ==============================<br><br><br>
';

//SOME NASTY SHIT
$err = 0; // no erros at begining :)
$unlinked = 0; //deleted at beggining ? ;]
$changed = 0; //changed files at beggining 

//ACCOUNTS CHECK
	echo '<b>'.$mode['accounts'].'</b><br><br>';

//number of digits
	gentime();		
		$files = glob($path['acc'] .'*.xml');
			foreach($files as $file) 
			{
					if(!in_array(parseFile($file,0), $except['acc']))
						{
				$xml = @simplexml_load_file($file);
		
			if(!$xml)
			{
				echo error_color(parseFile($file, 1).' '.$error['xml_syntax'].'<br>', red);
				++$err;
				
				if($unlink == 1)
					{
							if($UNIX == 1)
								{
									@chmod($file, 0777);
									}			
				$del = @unlink($file);
				
					if(!$del)
						{
							echo $error['deleting'];
							}
								else
								{
									echo error_color(parseFile($file, 1).' Deleted succesfully</br>', green);
									--$err;
									++$unlinked;
								}			
				}
				}
				
				$dig = strlen(trim(parseFile($file, 0)));
								
						if ($dig != $account['digits'] + 4)
							{
								echo error_color(parseFile($file, 1).' '.$error['wrong_digits'].'('.($dig -4).')<br>', red);
										++$err;
										
										if($unlink == 1)
					{
							if($UNIX == 1)
								{
									@chmod($file, 0777);
									}
				
				$del = @unlink($file);
				
					if(!$del)
						{
							echo $error['deleting'];
							}
								else
								{
									echo error_color(parseFile($file, 1).' Deleted succesfully</br>', green);
									--$err;
									++$unlinked;
								}
				
					}
							}
			}
			}
			
echo '* '.$title['digits'].' ['.round(gentime(), 4).' '.$sec.'.]<br><br>';	
	

//passwords

    gentime();

//$files = glob($path['acc'] . '*.xml');

foreach($files as $file) {
    $xml = @simplexml_load_file($file);
	if(!$xml)
			{
	//echo $excluded;

		if(!in_array(parseFile($file,0), $except['acc']))
						{
		
if(empty($xml['pass']))

    {
        echo error_color(parseFile($file, 1).' '.$error['empty_pass'].'<br>', red);
			++$err;
    }

						}
			}
						}
echo '* '.$title['pass'].' ['.round(gentime(), 4).' '.$sec.'.]<br><br>';

//types

    gentime(); 
    
       //$files = glob($path['acc'] . '*.xml');

foreach($files as $file) {
    $xml = @simplexml_load_file($file);
	
if(!in_array(parseFile($file,0), $except['acc']))
						{
if(empty($xml['type']) || $xml['type'] != 1)
    {
        echo error_color(parseFile($file, 1).' '.$error['wrong_type'].' ('.$xml['type'].')<br>', red);
			++$err;
			
			if($change == 1)
					{
				$xml['type'] = 1;   //change to type="1"
        $make_change = @file_put_contents($file, $xml->asXML());
		
						if(!$make_change)
						{
							echo $error['change'];
							}
								else
								{
									echo error_color(parseFile($file, 1).' Changed succesfully</br>', orange);
									--$err;
									++$changed;
									}
								
		}
    }

}
}
echo '* '.$title['types'].' ['.round(gentime(), 4).' '.$sec.'.]<br><br>';

//premdays

gentime();

 //$files = glob($path['acc'] . '*.xml');

foreach($files as $file) {
    
    $xml = @simplexml_load_file($file);
	
	if(!in_array(parseFile($file,0), $except['acc']))
						{
$prem =  trim($xml['premDays']); 
$lastsave = trim($xml['lastsaveday']);

	//string...lol
	//echo var_dump($prem);
    
if(($prem == ''))

    {
		
			echo error_color(parseFile($file, 1).' '.$error['prem_without_value'].'<br>', red);
				++$err;
    }	
        else if(is_numeric($prem) == false) 
            {
                
               echo error_color(parseFile($file, 1).' '.$error['prem_not_int'].'<br>', red); 
					++$err;
                
            }		
else if ($prem > $premDay['limit'])
	{
			echo error_color(parseFile($file, 1).' '.$error['prem_max'].' ('.$prem.')<br>', red);
				++$err;
				
					if($change == 1)
					{
				$xml['premDays'] = $premDay['rigid'];   //how many premdays will have
        $make_change = @file_put_contents($file, $xml->asXML());
		
						if(!$make_change)
						{
							echo $error['change'];
							}
								else
								{
									echo error_color(parseFile($file, 1).' Changed succesfully</br>', orange);
									--$err;
									++$changed;
									}					
		}

				}

else if ($prem < 0)
	{
		echo error_color(parseFile($file, 1).' '.$error['prem_less_zero'].' ('.$prem.')<br>', red);
			++$err;

				if($change == 1)
					{
				$xml['premDays'] = $premDay['rigid'];   //how many premdays will have
				
        $make_change = @file_put_contents($file, $xml->asXML());
		
						if(!$make_change)
						{
							echo $error['change'];
							}
								else
								{
									echo error_color(parseFile($file, 1).' Changed succesfully</br>', orange);
									--$err;
									++$changed;
									}
			}

}
}
}

echo '* '.$title['prem'].' ['.round(gentime(), 4).' '.$sec.'.]<br><br>'; 

gentime();

 //$files = glob($path['acc'] . '*.xml');

		foreach($files as $file) {
if(!in_array(parseFile($file,0), $except['acc']))
						{
		$cont = file_get_contents($file);
				$result = count(explode('character name', $cont));
				
				if($result < 2)
				
				{
					echo error_color(parseFile($file, 1).' '.$error['empty_account'].'<br>', red);
				++$err;								
										
			if($unlink == 1)
					{
							if($UNIX == 1)
								{
									@chmod($file, 0777);
									}
				
				$del = @unlink($file);
				
					if(!$del)
						{
							echo $error['deleting'];
							}
								else
								{
									echo error_color(parseFile($file, 1).' Deleted succesfully</br>', green);
									--$err;
									++$unlinked;
																		}
									}
								}
			
						}
						}
						
echo '* '.$title['char'].' ['.round(gentime(), 4).' '.$sec.'.]<br><br>';

//PLAYERS

echo '<b>'.$mode['players'].'</b><br><br>';

// acc number
gentime();

 $files = glob($path['pla'] . '*.xml');

foreach($files as $file) {
	if(!in_array(parseFile($file,0), $except['pla']))
	{
    
    $xml = @simplexml_load_file($file);
		
		if(!$xml)
			{
				echo error_color(parseFile($file, 1).' '.$error['xml_syntax'].'<br>', red);
				++$err;
				
				if($unlink == 1)
					{
							if($UNIX == 1)
								{
									@chmod($file, 0777);
									}
				
				$del = @unlink($file);
				
					if(!$del)
						{
							echo $error['deleting'];
							}
								else
								{
									echo error_color(parseFile($file, 1).' Deleted succesfully</br>', green);
									--$err;
									++$unlinked;
																		}
				
				}			
				}
				
			$a =  trim($xml['account']); 

	if(!file_exists($path['acc'].$a.'.xml'))
		{
				
			echo error_color(parseFile($file, 1).' '.$error['account_miss'].' <br>', red);
				++$err;
				
				if($unlink == 1)
					{
								if($UNIX == 1)
									{
										@chmod($file, 0777);
										}
				$del = @unlink($file);
				
					if(!$del)
						{
							echo $error['deleting'];
							}
								else
								{
									echo error_color(parseFile($file, 1).' Deleted succesfully</br>', green);
									--$err;
									++$unlinked;
																		}
									}
			
			}
			
}
}
echo ' * '.$title['acc_num'].' ['.round(gentime(), 4).' '.$sec.'.]<br><br>';

// health and health max
gentime();

 //$files = glob($path['pla'] . '*.xml');

foreach($files as $file) {
        
        if(!in_array(parseFile($file,0), $except['pla']))
        {
    
    $xml = @simplexml_load_file($file);
    
         $healthNow = (int)$xml->health['now'];
         $healthMax = (int)$xml->health['max'];
         
         if($healthNow < 0 || $healthMax < 0)
            {
                
                echo error_color(parseFile($file, 1).' '.$error['health'].' <br>', red);
                ++$err;
                
                if($unlink == 1)
                    {
                                if($UNIX == 1)
                                    {
                                        @chmod($file, 0777);
                                        }
                $del = @unlink($file);
                
                    if(!$del)
                        {
                            echo $error['deleting'];
                            }
                                else
                                {
                                    echo error_color(parseFile($file, 1).' Deleted succesfully</br>', green);
                                    --$err;
                                    ++$unlinked;
                                                                        }
                                    }

            }
    
        }
        
        
}

echo ' * '.$title['health'].' ['.round(gentime(), 4).' '.$sec.'.]<br><br>'; 

// mana & mana max
gentime();

 //$files = glob($path['pla'] . '*.xml');

foreach($files as $file) {
        
        if(!in_array(parseFile($file,0), $except['pla']))
        {
    
    $xml = @simplexml_load_file($file);
    
         $manaNow = (int)$xml->mana['now'];
         $manaMax = (int)$xml->mana['max'];
         
         if($manaNow < 0 || $manaMax < 0)
            {
                
                echo error_color(parseFile($file, 1).' '.$error['mana'].' <br>', red);
                ++$err;
                
                if($unlink == 1)
                    {
                                if($UNIX == 1)
                                    {
                                        @chmod($file, 0777);
                                        }
                $del = @unlink($file);
                
                    if(!$del)
                        {
                            echo $error['deleting'];
                            }
                                else
                                {
                                    echo error_color(parseFile($file, 1).' Deleted succesfully</br>', green);
                                    --$err;
                                    ++$unlinked;
                                                                        }
                                    }
             
                
            }
    
        }

}

echo ' * '.$title['mana'].' ['.round(gentime(), 4).' '.$sec.'.]<br><br>';   


// sex
gentime();

 //$files = glob($path['pla'] . '*.xml');

foreach($files as $file) {
		
		if(!in_array(parseFile($file,0), $except['pla']))
		{
    
    $xml = @simplexml_load_file($file);
		
   
			$s =  trim($xml['sex']); 
			//var_dump($s);
				if (!is_numeric($s))
					
					{
						echo error_color(parseFile($file, 1).' '.$error['sex_numeric'].'<br>', red);
						++$err;
						
						}
						
						if($s < 0 || $s > 4)
						{
							echo error_color(parseFile($file, 1).' '.$error['sex_value'].' ('.$s.')<br>', red);
						++$err;
						}
			}
			}
			
echo ' * '.$title['sex_dig'].' ['.round(gentime(), 4).' '.$sec.'.]<br><br>';


//lookdir

gentime();

foreach($files as $file) {
	if(!in_array(parseFile($file,0), $except['pla']))
	{
    
    $xml = @simplexml_load_file($file);
		
   
			$ldir =  trim($xml['lookdir']); 
				//var_dump($ldir);
			
							$ldir_allowed = array ("0","1","2","3");
					
							if(!in_array($ldir, $ldir_allowed))
							{
								echo error_color(parseFile($file, 1).' '.$error['lookdir_value'].' ('.$ldir.')<br>', red);
									++$err;
								
								}			
				}
				}
				

echo ' * '.$title['lookdir'].' ['.round(gentime(), 4).' '.$sec.'.]<br><br>';	

//exp

gentime();

foreach($files as $file) {
	if(!in_array(parseFile($file,0), $except['pla']))
	{
    
    $xml = @simplexml_load_file($file);
		
   
			$exp =  trim($xml['exp']); 
				
					if($exp < 0 || !is_numeric($exp))
					
						{
							echo error_color(parseFile($file, 1).' '.$error['exp_value'].'('.$exp.')<br>', red);
									++$err;
							
							}
				}
				}
					

echo ' * '.$title['exp'].' ['.round(gentime(), 4).' '.$sec.'.]<br><br>';


gentime();

foreach($files as $file) {
	if(!in_array(parseFile($file,0), $except['pla']))
	{
		$xml = simplexml_load_file($file);
	
		$pvoc = $xml['voc'];
		$pcap = $xml['cap'];
		$plevel = $xml['level'];
		$phmax = $xml->health['max'];
		$pmmax = $xml->mana['max'];

	#capacity vs level

		if($pvoc == 1) //sorc
		{
		$trueCap = (($plevel -1)*10)+300;
			if($trueCap != $pcap)
			{
			echo error_color(parseFile($file, 1).' '.$error['cap_l'].' should have: '.$trueCap.' now - '.$pcap.'<br>', red);
									++$err;

			}
		}


		if($pvoc == 2) //druid
		{
		$trueCap = (($plevel-1)*10)+300;
			if($trueCap != $pcap)
			{
			echo error_color(parseFile($file, 1).' '.$error['cap_l'].' should have: '.$trueCap.' now - '.$pcap.'<br>', red);
									++$err;

			}
		}


		if($pvoc == 3) //palladin
		{
		$trueCap = (($plevel-1)*20)+300;
			if($trueCap != $pcap)
			{
			echo error_color(parseFile($file, 1).' '.$error['cap_l'].' should have: '.$trueCap.' now - '.$pcap.'<br>', red);
									++$err;

			}
		}

		if($pvoc == 4) //knight
		{
		$trueCap = (($plevel-1)*25)+300;
			if($trueCap != $pcap)
			{
			echo error_color(parseFile($file, 1).' '.$error['cap_l'].' should have: '.$trueCap.' now - '.$pcap.'<br>', red);
									++$err;

			}
		}
		
#mana gain vs level

if($pvoc == 1) //sorc
		{
		$trueMana = (($plevel-1)*30);
			if($trueMana != $pmmax)
			{
			echo error_color(parseFile($file, 1).' '.$error['manamax_l'].' should have: '.$trueMana.' now - '.$pmmax.'<br>', red);
									++$err;

			}
		}

		if($pvoc == 2) //druid
		{
		$trueMana = (($plevel-1)*30);
			if($trueMana != $pmmax)
			{
			echo error_color(parseFile($file, 1).' '.$error['manamax_l'].' should have: '.$trueMana.' now - '.$pmmax.'<br>', red);
									++$err;
			}
		}

	if($pvoc == 3) //palladin
		{
		$trueMana = (($plevel-1)*15);
			if($trueMana != $pmmax)
			{
			echo error_color(parseFile($file, 1).' '.$error['manamax_l'].' should have: '.$trueMana.' now - '.$pmmax.'<br>', red);
									++$err;

			}
		}

		if($pvoc == 4) //knight
		{
		$trueMana = (($plevel-1)*5);
			if($trueMana != $pmmax)
			{
			echo error_color(parseFile($file, 1).' '.$error['manamax_l'].' should have: '.$trueMana.' now - '.$pmmax.'<br>', red);
									++$err;
			}
		}
	#health vs level

if($pvoc == 1) //sorc
		{
		$trueHealth = (($plevel-1)*5)+150;
			if($trueHealth != $phmax)
			{
			echo error_color(parseFile($file, 1).' '.$error['healthmax_l'].' should have: '.$trueHealth.' now - '.$phmax.'<br>', red);
									++$err;
			}
		}
		if($pvoc == 2) //druid
		{
		$trueHealth = (($plevel-1)*5)+150;
			if($trueHealth != $phmax)
			{
			echo error_color(parseFile($file, 1).' '.$error['healthmax_l'].' should have: '.$trueHealth.' now - '.$phmax.'<br>', red);
									++$err;

			}
		}

	if($pvoc == 3) //palladin
		{
		$trueHealth = (($plevel-1)*10)+150;
			if($trueHealth != $phmax)
			{
			echo error_color(parseFile($file, 1).' '.$error['healthmax_l'].' should have: '.$trueHealth.' now - '.$phmax.'<br>', red);
									++$err;
			}
		}


		if($pvoc == 4) //knight
		{
		$trueHealth = (($plevel-1)*15)+150;
			if($trueHealth != $phmax)
			{
			echo error_color(parseFile($file, 1).' '.$error['healthmax_l'].' should have: '.$trueHealth.' now - '.$phmax.'<br>', red);
									++$err;

			}
		}

	}
}


echo ' * '.$title['var_l'].' ['.round(gentime(), 4).' '.$sec.'.]<br><br>';

//VIP ACCOUNTS

gentime();

$files = glob($path['vip'] . '*.xml');
foreach($files as $file) {
	if(!in_array(parseFile($file,0), $except['acc']))
	{

	 $xml = @simplexml_load_file($file);	
 			if(!$xml)
			{
				echo error_color(parseFile($file, 1).' '.$error['xml_syntax'].'<br>', red);
				++$err;
				
				if($unlink == 1)
					{
						if($UNIX == 1)
							{
								@chmod($file, 0777);
							}
				
				$del = @unlink($file);
				
					if(!$del)
						{
							echo $error['deleting'];
							}
								else
								{
									echo error_color(parseFile($file, 1).' Deleted succesfully</br>', green);
									--$err;
									++$unlinked;
																		}				
				}
				
				}
			if(!file_exists($path['acc'].parseFile($file, 0)))
					{
					echo error_color(parseFile($file, 1).' '.$error['vip_exists'].'<br>', red);
						++$err;
						if($unlink == 1)
					{
						if($UNIX == 1)
							{
								@chmod($file, 0777);
							}			
				$del = @unlink($file);			
					if(!$del)
						{
							echo $error['deleting'];
							}
								else
								{
									echo error_color(parseFile($file, 1).' Deleted succesfully</br>', green);
									--$err;
									++$unlinked;
																		}				
				}

					}					
				}
				}
					

echo ' * '.$title['vip'].' ['.round(gentime(), 4).' '.$sec.'.]<br><br>';		

echo $error['sum'].' ';
	listErrors();
		echo '</br>';
		
		if($unlink == 1)
				{
					echo $error['unlinked'].' ';
							deletedFiles();
								echo '</br>';
				}
					
					if($change == 1)
						{
							echo $error['changed'].' ';
							changedFiles();
							}				
				if($saveLog == 1 || $saveLog == 2)
					{
				
				$output .= '<html>  
<head> 
<!-- Oink! oink! -->
<!-- XML Server Validator by Pavlus -->
<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>
<title>Validator XML - Raport</title>
<body> ';
				
	$output .= ob_get_contents(); //take cached content
	
		$output .='</body>  
</html>  ';
	
		// Raport saved to file
		
		$date_rap = date("F_j_Y_g_i_a");
		
			$raport_file = './LOG/Raport_'.$date_rap.'.html';
			    
			
					$fp = @fopen($raport_file, "a+"); 

						$fw = @fwrite($fp, $output);  //save													
							@fclose($fp); 
							
							if($saveLog == 2)
							    {							
							ob_end_clean();
							    //cache cleaner
							}						
					}				
		$set_last_timeout = @ini_set('max_execution_time',$timeout); //change exec.time to default value
		if($UNIX == 1)
			{
				//lets get back to old chmod we assume that was the same for players
				//and account folder :)
			@chmod($path['pla'], $old_chmod_changed); 
				@chmod($path['acc'], $old_chmod_changed);
					@chmod($path['vip'], $old_chmod_changed);
				}
		


?>
